pkgs/net-lib/net/server.rkt

```diff
@@ -0,0 +1,106 @@
+#lang racket/base
+
+(require racket/contract
+         racket/tcp)
+
+(provide
+ (contract-out
+  [start-server (server-proc/c (-> void?))]
+  [run-server (server-proc/c void?)]))
+
+(define (server-proc/c res/c)
+  (->* ((or/c #f string?)
+        listen-port-number?
+        (-> input-port? output-port? any))
+       (#:reuse? any/c
+        #:max-allow-wait exact-nonnegative-integer?
+        #:max-concurrent (or/c +inf.0 exact-positive-integer?)
+        #:listen-proc (-> listen-port-number? exact-nonnegative-integer? any/c (or/c #f string?) evt?)
+        #:accept-proc (-> any/c (values input-port? output-port?))
+        #:close-proc (-> any/c void?)
+        #:timeout-evt-proc (-> thread? input-port? output-port? boolean? evt?))
+       res/c))
+
+(define (start-server host port handle
+                      #:reuse? [reuse? #t]
+                      #:max-allow-wait [max-allow-wait 4]
+                      #:max-concurrent [max-concurrent +inf.0]
+                      #:listen-proc [listen tcp-listen]
+                      #:accept-proc [accept tcp-accept]
+                      #:close-proc [close tcp-close]
+                      #:timeout-evt-proc [make-timeout-evt (λ (_thd _in _out _break-sent?) never-evt)])
+  (define can-break?
+    (break-enabled))
+  (define paramz
+    (current-parameterization))
+  (define listener
+    (listen port max-allow-wait reuse? host))
+  (define server-thd
+    (thread
+     (lambda ()
+       (dynamic-wind
+         void
+         (lambda ()
+           (with-handlers ([exn:break? void])
+             (let loop ([in-progress 0])
+               (loop
+                (with-handlers ([exn:fail:network?
+                                 (λ (e)
+                                   (begin0 in-progress
+                                     ((error-display-handler)
+                                      (format "Connection error: ~a" (exn-message e))
+                                      e)))])
```

Comment:
This seems like a place where it would be nice to allow interposition, though I'm not sure exactly how.